### ISSUE TITLE
unused-command-line-argument warning with Apple Clang

### ISSUE THREAD

**Issue by StrikerRUS** [microsoft/LightGBM#1805]
There is a warning

```
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
```

in the current `master` branch.

I've done binsearching among [available Xcode versions on Azure Pipelines](https://github.com/Microsoft/azure-pipelines-image-generation/blob/master/images/macos/macos-Readme.md#xcode) and found that the lowest Xcode version which recognizes `-pthread` and doesn't produce the warning is `9.3`.

ping @guolinke and @chivee

**Comment by guolinke** [comment]
@StrikerRUS refer to https://stackoverflow.com/questions/2391194/what-is-gs-pthread-equiv-in-clang/19382746#19382746

**Comment by chivee** [comment]
maybe just adding ` -Qunused-arguments option` for those under 9.3

**Comment by StrikerRUS** [comment]
@chivee Won't this hide from us future (possibly dangerous) warnings?

I've tried the following solution, but unfortunately warning hasn't gone.
```
if((CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang") AND (CMAKE_CXX_COMPILER_VERSION VERSION_LESS "9.1.0"))
    if("${CMAKE_SHARED_LINKER_FLAGS}" STREQUAL "")
        string(REPLACE "-pthread" "" OLD_APPLE_CLANG_SHARED_LINKER_FLAGS "${CMAKE_CXX_FLAGS}")
    else()
        string(REPLACE "-pthread" "" OLD_APPLE_CLANG_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS}")
    endif()
    set(CMAKE_SHARED_LINKER_FLAGS "${OLD_APPLE_CLANG_SHARED_LINKER_FLAGS}")
    if("${CMAKE_EXE_LINKER_FLAGS}" STREQUAL "")
        string(REPLACE "-pthread" "" OLD_APPLE_CLANG_EXE_LINKER_FLAGS "${CMAKE_CXX_FLAGS}")
    else()
        string(REPLACE "-pthread" "" OLD_APPLE_CLANG_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS}")
    endif()
    set(CMAKE_EXE_LINKER_FLAGS "${OLD_APPLE_CLANG_EXE_LINKER_FLAGS}")
endif()
```

**Comment by StrikerRUS** [comment]
@guolinke I've tried: https://github.com/Microsoft/LightGBM/issues/1805#issuecomment-440825017. Unfortunately, without any luck 😢 .

**Comment by StrikerRUS** [comment]
Closed in favor of being in #2302. We decided to keep all feature requests in one place.

Welcome to contribute this feature! Please re-open this issue (or post a comment if you are not a topic starter) if you are actively working on implementing this feature.

**Comment by StrikerRUS** [comment]
> I've done binsearching among [available Xcode versions on Azure Pipelines](https://github.com/Microsoft/azure-pipelines-image-generation/blob/master/images/macos/macos-Readme.md#xcode) and found that the lowest Xcode version which recognizes `-pthread` and doesn't produce the warning is `9.3`.

According to the 
- > the lowest Xcode version which recognizes `-pthread` and doesn't produce the warning is `9.3`;
- we have no easy access to Xcode versions less than `10.3` (#4797);
- this is just warning, not error

I'm closing this issue.
